File: src/logic/EventBuffer.ts
// Compact, single-buffer event stream utilities for JS<->WASM turns.
// Layout per event (all little-endian):
// - header u32: low 16 bits = type, high 16 bits = size in dwords (including header)
// - payload: packed words; floats read via Float32 view.

export const EVENT_BUFFER_WORDS = 65536;

export class EventBuffer {
  u32!: Uint32Array;
  f32!: Float32Array;
  capWords: number = 0;
  cursor: number = 0;

  constructor(buffer: ArrayBuffer, basePtrBytes: number, capWords: number) {
  this.u32 = new Uint32Array(buffer, basePtrBytes, capWords);
  this.f32 = new Float32Array(buffer, basePtrBytes, capWords);
  this.capWords = capWords;
  this.cursor = 0;
  }

  public writeHeader(type: number, sizeWords: number): void {
  if (this.cursor + sizeWords >= this.capWords) throw new Error("event buffer full");
  this.u32[this.cursor] = ((sizeWords & 0xffff) << 16) | (type & 0xffff);
  }

  public beginFrame(): void {
  this.cursor = 0;
  }

  public commitFrame(): void {
  this.u32[this.cursor] = 0;
  }
}
/*
export function forEachEvent(
  buf: EventBuffer,
  handlers: {
  mixed?: (x: number, y: number) => void;
  unknown?: (type: number, sizeWords: number, startWord: number) => void;
  },
): void {
  let p = 0;
  const cap = buf.capWords;
  while (true) {
  const header = buf.u32[p] >>> 0;
  if (header === 0) break;
  const type = header & 0xffff;
  const size = header >>> 16;
  if (size <= 0 || p + size > cap) break; // malformed; stop

  switch (type) {
    case EVT_MIXED: {
    const x = buf.f32[p + 1];
    const y = buf.f32[p + 2];
    handlers.mixed?.(x, y);
    break;
    }
    default:
    handlers.unknown?.(type, size, p);
    break;
  }

  p += size;
  }
}
*/

File: src/wasm/event_buffer.cpp
#include "event_buffer.h"
#include <emscripten/emscripten.h>
#include "wasm_log.h"

EventBuffer g_event_buffer;

void EventBuffer::set(uint8_t* base, uint32_t cap_words_) {
  uintptr_t addr = reinterpret_cast<uintptr_t>(base);
  addr = (addr + 3) & ~static_cast<uintptr_t>(3);
  u32_base = reinterpret_cast<uint32_t*>(addr);
  f32_base = reinterpret_cast<float*>(addr);
  cap_words = cap_words_;
  cursor = 0u;
}

void EventBuffer::begin_frame() {
  cursor = 0u;
}

void EventBuffer::commit_frame() {
  if (u32_base && cursor < cap_words) u32_base[cursor] = 0u;
}

void EventBuffer::write_header(uint16_t type, uint16_t size_words) {
  if (cursor + size_words >= cap_words){
    wasm_console_error("Event buffer full");
    return;
  }
  u32_base[cursor] = (static_cast<uint32_t>(size_words) << 16) | static_cast<uint32_t>(type);
  cursor += size_words;
}
